### PR TITLE
GH-2675 add new RepositoryConnection.prepare method

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #2675  

Briefly describe the changes proposed in this PR:

- added RepositoryConnection.prepare and implementations in core classes
- added handling over remote protocol
- added tests, including a SHACL-specific use case test

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits down to one or a few meaningful commits
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

### CODE REVIEW

File: core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepositoryConnection.java [hunk 1]
```diff
@@ -294,6 +294,23 @@ public long size(Resource... contexts) throws RepositoryException {
 		}
 	}
 
+	@Override
+	public void prepare() throws RepositoryException {
+		if (this.getRepository().getServerProtocolVersion() < 12) {
+			// Action.PREPARE is not supported in Servers using protocols older than version 12.
```
Comment (hmottestad) [inline core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepositoryConnection.java:300]: Could you log an error message here?

File: core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java [hunk 2]
```diff
@@ -405,6 +404,11 @@ public TupleQuery prepareTupleQuery(QueryLanguage ql, String query, String base)
 		throw new UnsupportedQueryLanguageException("Unsupported query language " + ql);
 	}
 
+	@Override
+	public void prepare() throws RepositoryException {
+		// no-op, not supported in SPARQL protocol
```
Comment (hmottestad) [inline core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java:409]: I would prefer an unsupported operation exception, so users don’t spend time trying to figure out why prepare isn’t doing anything.

**Comment by hmottestad** [comment]
Just minor comments on my part. Lgtm.

File: core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java:409]: Yeah good point, I'll do the same for the default method.

**Pull request merged**